pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -58,41 +41,88 @@ func (res *claimInfo) addPodReference(podUID types.UID) {
 	res.Lock()
 	defer res.Unlock()
 
-	res.podUIDs.Insert(string(podUID))
+	res.PodUIDs.Insert(string(podUID))
 }
 
 func (res *claimInfo) deletePodReference(podUID types.UID) {
 	res.Lock()
 	defer res.Unlock()
 
-	res.podUIDs.Delete(string(podUID))
+	res.PodUIDs.Delete(string(podUID))
 }
 
 // claimInfoCache is a cache of processed resource claims keyed by namespace + claim name.
 type claimInfoCache struct {
 	sync.RWMutex
+	state     state.CheckpointState
 	claimInfo map[string]*claimInfo
 }
 
+func newClaimInfo(driverName string, claimUID types.UID, claimName, namespace string, podUIDs sets.Set[string], cdiDevice []string) (*claimInfo, error) {
+
+	// NOTE: Passing CDI device names as annotations is a temporary solution
+	// It will be removed after all runtimes are updated
+	// to get CDI device names from the ContainerConfig.CDIDevices field
+	annotations, err := generateCDIAnnotations(claimUID, driverName, cdiDevice)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate container annotations, err: %+v", err)
+	}
+	claimInfoState := state.ClaimInfoState{
+		DriverName: driverName,
+		ClaimUID:   claimUID,
+		ClaimName:  claimName,
+		Namespace:  namespace,
+		PodUIDs:    podUIDs,
+		CdiDevices: cdiDevice,
+	}
+	claimInfo := claimInfo{
+		ClaimInfoState: claimInfoState,
+		annotations:    annotations,
+	}
+	return &claimInfo, nil
+}
+
 // newClaimInfoCache is a function that returns an instance of the claimInfoCache.
-func newClaimInfoCache() *claimInfoCache {
-	return &claimInfoCache{
+func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error) {
+
+	stateImpl, err := state.NewCheckpointState(stateDir, checkpointName)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize checkpoint manager, please drain node and remove dra state file, err: %+v", err)
+	}
+
+	curState, err := stateImpl.GetOrCreate()
+	if err != nil {
+		return nil, fmt.Errorf("error calling GetOrCreate() on checkpoint state: %v", err)
+	}
+
+	cache := &claimInfoCache{
+		state:     stateImpl,
 		claimInfo: make(map[string]*claimInfo),
 	}
+
+	for _, entry := range curState {
+		info, err := newClaimInfo(
+			entry.DriverName,
+			entry.ClaimUID,
+			entry.ClaimName,
+			entry.Namespace,
+			entry.PodUIDs,
+			entry.CdiDevices,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create claimInfo %+v: %+v", info, err)
+		}
+		cache.add(info)
+	}
+
+	return cache, nil
 }
 
-func (cache *claimInfoCache) add(claim, namespace string, res *claimInfo) error {
+func (cache *claimInfoCache) add(res *claimInfo) {
 	cache.Lock()
 	defer cache.Unlock()
 
-	key := claim + namespace
-	if _, ok := cache.claimInfo[key]; ok {
-		return fmt.Errorf("claim %s, namespace %s already cached", claim, namespace)
-	}
```

Comment:
@moshe010 can you explain why this check was removed?

Reply:
conceptually panic in golang is an assert without condition, but I'm not sure we should panic here. this code is called when we prepare resources and that operation is indempotent, i.e. we might have the same or at least functionally the same result.

Reply:
But it should never reach the  `add()` call because it is protected earlier by the call to `get()`.

Reply:
In fact if it ever did reach the `add()` call a second time with the same results, it would error out because of the check that was previously there ...

Reply:
> But it should never reach the add() call because it is protected earlier by the call to get()

True, but it's done on the caller side, so can be changed there. Checking that we're not rewriting data still makes sense to me.

Reply:
I guess it comes down to how you view the API. To me, I think of `delete()`, `get()` and `add()` as all being idempotent and should explicitly not do any checking. They should blindly do what is told of them (and if you wanted to do any checking, this should happen outside of these calls -- which is in fact what we are currently doing).

Reply:
My understanding was that `add` should add, but not rewrite. That was a reason to check it. You're right, it was an assertion, something not expected, but potentially dangerous and therefore worth checking.

Reply:
If we keep the check, then I would argue we also need to introduce a check on `delete()` through the same reasoning. Though I would still argue it is not necessary from my point of view in either call.

Reply:
I'm ok with this. Just thought it would be harder to debug silent cache rewrites. As for `delete` call - it's harmless to attempt deleting something that was already deleted.